### PR TITLE
Application crd

### LINKED ISSUE #1106: Application Custom Resource for Kubeflow Deployments
We should consider creating an [Application Resource](https://github.com/kubernetes-sigs/application) to describe a Kubeflow deployment.

At this point the Application Resource mostly provides meta information.

I think the immediate use cases for it would be

1. Use spec.assemblyPhase to report phase of the Kubeflow deployment
1. We could use the Application as the resource to associate with Kubernetes events related to the Kubeflow deployment.

Related issue #955

### PR DESCRIPTION
**Pull request by kkasravi**
fixes #1106

@jlewi early view of `kind: Application` where the Application CR is being generated.
The [README.md](https://github.com/kubeflow/kubeflow/pull/1633/files#diff-c6c18b175e7f985f71f989f573c161f8) shows how to generate the CR.
TODO: finish the controller sync hook (uses metacontroller) which will need to provide the status of the Application Kind.

An option is to have the sync hook create all the component resources. This could be auto-generated in application.libsonnet since application.libsonnet generates the sync hook and has access to all of the components. This would setup ownership of all component resources as being the Application Controller. I can do a demo next week if you would like - possibly to community if you think it is a direction we would like to entertain.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/1633)

### CODE REVIEW

**Comment by jlewi** [comment]
This looks great.

**Comment by jlewi** [comment]
Either in this PR or as a follow on issue it might be nice to provide version information; e.g. Kubeflow v0.3.0

**Comment by kkasravi** [comment]
@jlewi remaining:
- [ ] applicationController requires cluster RBAC to deploy that i'm fixing.
- [ ] Metacontroller should be separated out into its own package that requires cluster-admin RBAC to deploy and is part of admin setup. Related to first issue

Feel free to review. I've noted your request to make the application-controller-hooks configmap more pluggable so other options could be used as well

**Comment by jlewi** [comment]
Can you remove the hold from the PR and assign it to me when if its ready for review?

File: kubeflow/core/prototypes/prometheus.jsonnet [hunk 1]
```diff
@@ -3,9 +3,9 @@
 // @description Provides prometheus prototype in kubeflow core.
 // @shortDescription Prometheus Service.
 // @param name string Name for the component
-// @param projectId string GCP project id.
-// @param clusterName string GKE cluster name.
-// @param zone string GKE cluster zone.
+// @optionalParam projectId string GCP project id.
```
Comment (jlewi) [inline kubeflow/core/prototypes/prometheus.jsonnet]: Why this change? Could/Should this be factored out into a separate PR?

File: kubeflow/core/tests/tf-job_test.jsonnet [hunk 2]
```diff
@@ -11,7 +11,7 @@ local paramsv1alpha2 = {
   name:: "tf-job-operator",
   tfJobImage:: "gcr.io/kubeflow-images-public/tf_operator:v0.3.0",
   tfDefaultImage:: "null",
-  deploymentScope:: "cluster",
+  deploymentScope:: "namespace",
```
Comment (jlewi) [inline kubeflow/core/tests/tf-job_test.jsonnet]: What is this change for?

File: scripts/kfctl.sh [hunk 3]
```diff
@@ -235,13 +235,14 @@ ksApply () {
...
 
   # Create all the core components
-  ks apply default -c ambassador
-  ks apply default -c jupyterhub
-  ks apply default -c centraldashboard
-  ks apply default -c tf-job-operator
-  ks apply default -c argo
-  ks apply default -c katib
-  ks apply default -c spartakus
+  # ks apply default -c ambassador
```
Comment (jlewi) [inline scripts/kfctl.sh]: So you're changing the default deployment mechanism to deploy via the metacontroller from in cluster?
Why not leave the default as deploying client side?

**Comment by kkasravi** [comment]
thanks @jlewi for the review. 
- [x] Removed extraneous changes
- [x] Corrections to katib, argo that have namespace when deploying ClusterRole, ClusterRoleBinding
- [x] Only build the Application kind based on components being deployed
- [x] Do not deploy incluster

File: kubeflow/application/prototypes/application.jsonnet [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// @apiVersion 0.1
+// @name io.ksonnet.pkg.application
+// @description application Component
+// @shortDescription application Component
+// @param name string Name
+// @optionalParam type string kubeflow Type of application.
+// @optionalParam version string 0.3 Version of kubeflow
+// @optionalParam namespace string null Namespace to use for the components. It is automatically inherited from the environment if not set.
+// @optionalParam bootstrap string false install cluster-level resources for application
```
Comment (jlewi) [inline kubeflow/application/prototypes/application.jsonnet]: What is the bootstrap parameter for?

File: kubeflow/application/application.libsonnet [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+          } else {
+            kind: resource.kind,
+          },
+          resource,
+        ],
+      },
+      return:: groupKindAndResource,
+    }.return,
+
+    local perComponent(name) = {
```
[the remark this answers is not in the document]
Comment (kkasravi) [reply kubeflow/application/application.libsonnet]: no - it's the same list - just filtered. The bootstrap=true parameter will return resources that do not have a namespace.

File: scripts/kfctl.sh [hunk 6]
```diff
@@ -225,14 +225,13 @@ ksApply () {
...
-  ks apply default -c ambassador
-  ks apply default -c jupyterhub
-  ks apply default -c centraldashboard
-  ks apply default -c tf-job-operator
-  ks apply default -c argo
-  ks apply default -c katib
-  ks apply default -c spartakus
+  # Create the kubeflow application
+  # - first create all cluster resources
+  ks param set application bootstrap true
```
[the remark this answers is not in the document]
Comment (kkasravi) [reply scripts/kfctl.sh]: i agree - there should be a check that all cluster resources were deployed. 

The concept is to separate the 'devops' part from the data scientist part.
That is - different data scientists could now deploy the application into 
different namespaces. Or different data scientists could share one namespace where 
the application is deployed by the 'owner'. We have a controller implementation that 
creates a namespace on behalf of a user. We've shown under RBAC rules that we use that 
each data scientists can not 'see' other data scientist's namespaces, jobs.
I can open this PR for review, though it is similar to #1769.

**Comment by kkasravi** [comment]
@jlewi 
> Could we narrow the focus of this PR to just creating the Application CRD to provide meta info about the Kubeflow Deployment and in a separate PR create a meteacontroller to roll out Kubeflow from a ConfigMap or PD containing a ksonnet app.

I may have misunderstood what you were thinking of. This PR was modified to create all the resources client-side as well as the application CR that contains the resources created. However - as noted above - it first filters the resources looking for cluster level resources. This must be executed by 'devops' or someone with a clusterrolebinding that has a clusterrole that allows create, get, list, update, patch, delete, watch on CustomResourceDefinition, ClusterRole, ClusterRoleBinding.

File: kubeflow/application/prototypes/application.jsonnet (hunk 4 above)
Comment (kkasravi) [reply kubeflow/application/prototypes/application.jsonnet]: Described below. A different name might be used - maybe scope={Cluster, Namespaced} - similar to CRDs.

File: scripts/kfctl.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (kkasravi) [reply scripts/kfctl.sh]: > My suggestion would be to scope this PR to just providing an application resource whose sole purpose is to provide meta information about the Kubeflow deployment. This is high value because it will provide users visibility into Kubeflow.

Got it. So just generate the Application Kind based on what components have been generated, with an optional param if you want to generate an application for some of the components. `ks apply default -c application` would result in 
```yaml
all: [
  self.applicationCRD,
  self.application,
],
```
The application would hold information about all kf_app/ks_app/components unless specific components were provided eg `ks param set application components '["ambassador", "jupyterhub"]'`

[the remark this answers is not in the document]
Comment (kkasravi) [reply scripts/kfctl.sh]: > So the infra team will want to deploy Kubeflow such that each user gets their own namespace. So within that namespace the user can launch their notebooks, models, jobs.

The user isolation mechanism we've prototyped has some overlap. 

If I understand correctly - we could have several global metacontrollers - notebook-controller, model-controller, job-controller. These metacontrollers are shared across teams, users. Each one would take a CRD from a user (perhaps emitted from a ui) and generate the child resources required into the namespace owned|shared by the user. For notebooks the notebook-controller would create child resources {service, deployment} where the container would be the notebook image. Would a separate application be instantiated per metacontroller and dependencies defined so there is an overall kubeflow-controller that is the ownerReference for everything?

[the remark this answers is not in the document]
Comment (jlewi) [reply scripts/kfctl.sh]: Correct. 

If you look at the controllers we have today

- JupyterHub
- Argo
- TFJob
- SeldonDeployment

These can be shared across users/teams. Each user/team can create their own resources Jupyter notebooks, Argo workflows, TFJobs, SeldonDeployments in their own namespace but share the controller.

I believe this has how most K8s built in controllers (e.g. ReplicaSet, Job controller work).

In some cases we have additional work to do to support multi-user. For example, in Jupyter I think we want to use ISTIO to restrict web access to each notebook so that only user A can access user A's notebook.

**Comment by jlewi** [comment]
I think the only part in this PR that I'm confused about is why you replaced the liness

````
ks apply default -c cloud-endpoints	  # Deploy the kubeflow application
  ks apply default -c cert-manager	  # - first create all cluster resources
  ks apply default -c iap-ingress	  ks param set application components '["cloud-endpoints", "cert-manager", "iap-ingress"]' 
  ks apply default -c pytorch-operator
````

with 

```
  # Deploy the kubeflow application
  # - first create all cluster resources
  ks param set application bootstrap true
  ks apply default -c application
  # - second create all namespace scoped resources.
  ks param set application bootstrap false
  ks apply default -c application
```

Is this left over from an earlier draft and needs to be changed back? Unless I'm missing something it looks like this code is using the Application CRD to do the deployment and not doing the deployment client side by calling ks apply on the components.

**Comment by kkasravi** [comment]
> Is this left over from an earlier draft and needs to be changed back? Unless I'm missing something it looks like this code is using the Application CRD to do the deployment and not doing the deployment client side by calling ks apply on the components.

yes, absolutely, fixing this ...

**Comment by kkasravi** [comment]
@jlewi 

This is finally ready. It does have some ancillary changes in argo.libsonnet and others where I've removed namespace from things like ClusterRole, ClusterRoleBinding. I could do this in a separate PR if you would like. Per the changes, it's no longer necessary to filter resources by namespace. One thing we're not doing is adding an [annotation](https://github.com/kow3ns/community/blob/8cb87419883197032f4e5cce8d5518c9c5792f6c/keps/sig-apps/0003-kubernetes-application-api.md#discovering-an-objects-application) to each resource - which I had been doing when the application component deployed all the resources. Having the application component add the annotation and patch all resources could be error-prone but would be one option. The other option would be to have the controller do this. I'm not sure if that's exactly per what spec outlines but it does discuss [adoption](https://github.com/kow3ns/community/blob/8cb87419883197032f4e5cce8d5518c9c5792f6c/keps/sig-apps/0003-kubernetes-application-api.md#adopting-existing-components) of components

**Comment by jlewi** [comment]
This is great. Thank you so much.

/lgtm
/approve
